### DISCUSSION TITLE
Is there anyone know how to use twind with Nuxt?

### DISCUSSION THREAD

**Discussion by rockmandash** [tw-in-js/twind#182] [Q&A] [answered]
I'm starting a project with Nuxt for SEO purposes.

Currently, there are several options for me:

1. [Tailwind with JIT](https://tailwindcss.nuxtjs.org/)

   I am currently using this package right now

   This package works really well, and it supports `@nuxt/storybook` (the storybook for Nuxt) as well

   The downside is no `override:` variant which is __crucial__ to me, I make a lot of presentational components, the component composition that kind of stuff.

   I've tried to use [`addVariant`](https://tailwindcss.com/docs/plugins#adding-variants) the Tailwind provided, but it's too hard for me, you will need to understand PostCSS and walk the tree to get the desired outcome.

2. [WindiCSS](https://github.com/windicss/nuxt-windicss-module)

   Really great package and it supports Nuxt as well.

   But currently, no `override:` variant for WindiCSS as well.

   I've read the [windicss nuxt module source code](https://github.com/windicss/nuxt-windicss-module/blob/master/src/index.ts), but still don't understand how it works, it seems like there is a WebPack plugin involved.

3. [Twind Shim](https://twind.dev/handbook/the-shim.html)

   Twind Shim is my best way to go, however, there are three reasons holding me back from using it.

   1. My project is using FontAwesome, it comes with its own class name(`fa`) if I use twind.shim, it will output lots of warnings about the unknown rules in the console, I don't know if it's OK to ignore those warnings 😂, I know I can silence those temporarily.

   2. Prevent FOUC, the `<html lang="en" hidden>` might kill my SEO because lots of articles are saying search engines will just ignore those invisible elements.

   3. If I don't add `hidden` property, I will have FOUC 😂.

Is there anyone who has a working example combining Nuxt and Twind, I'm using pre-render mode, a temporary hack will be fine for me. 🙏

**Comment by sastan** [comment, marked as answer]
I hope this will help you:

1. There was/is a plan to have an `ignorelist` for warnings. For now I would just go with silent mode.
2. For the FOUC:
   1. a nuxt plugin like we have for next – I'm happy to help here, but I know almost nothing about nuxt.
   2. body style + class – `<body style="opacity: 0" class="!opacity-1">` or `<body style="display: hidden" class="!block">`
   3. umd with observe – start shimming as soon as possible
     
      ```html
      <head>
        <!--- these could all be combined into one script --->
        <!--- see this gist for an example: https://gist.github.com/sastan/c79d341b4bd2d1ec2096c437810b4f50 -->
        <script src="twind/twind.umd.js"></script>
        <script src="twind/observe/observe.umd.js"></script>
        <script>
          // Setup as needed
          twind.setup({})
          twindObserver.observe(document.element)
        </script>
      </head>
      ```

Please let me know if I can help in any way.

**Reply by rockmandash** [reply]
Wow, thank you for your fast reply!

It definitely helped me!

If the warning can be ignored, I might go with the shim solution!

__start shimming as soon as possible__ this is too cool, I will try if I can get the shimming start faster!

I have just tried `twind-cli` to generate a static CSS file for Nuxt to hot reload. But haven't succeeded yet!

I was thinking if any SSR framework has hot reload capability, the `twind-cli` might be a simple generic integration.

**Reply by tylerforesthauser** [reply]
Hang tight and I will find and tune up my codesandbox to show you an integration I feel works pretty well. It integrates Twind as a globally injected function you can access with `$tw`. 

```
<template>
  <div :class="$tw('bg-black')">
```

It's a configurable module and works with SSR, SSG and SPA.

**Reply by rockmandash** [reply]
@tylerforesthauser OMG can't wait to see it!

**Reply by tylerforesthauser** [reply]
Alrighty, here we go. 

[Codesandbox](https://codesandbox.io/s/twind-nuxtjs-static-deployment-zy02b)
[GitHub Repo](https://github.com/tylerforesthauser/twind-nuxt)

The main things you'll want to look at are `/modules/twind/` and the `nuxt.config.js`. If you have any questions, suggestions, concerns, whatever, please let me know!

**Reply by sastan** [reply]
This looks great! Could that be refactored into a plugin within the use-with-twind repo?

**Reply by rockmandash** [reply]
@tylerforesthauser This is insane, I just test it out both on development and pre-rendered HTML on my project and it actually works!

And it works with `@nuxt/storybook` as well, I just tried.

This is totally a game-changer! 🤯

**Reply by tylerforesthauser** [reply]
> This looks great! Could that be refactored into a plugin within the use-with-twind repo?

@sastan Absolutely! I will submit a PR as soon as I can. 👍
